proposals/Symbiosis Protocol.md

```diff
@@ -0,0 +1,393 @@
+Good project scope aims for ~3-5 engineers for 1-3 months (though feel free to suggest larger-scoped projects anyway). Projects do not include regular day-to-day maintenance and improvement work, e.g. on testing, tooling, validation, code clarity, refactors for future capability, etc.
+-->
+
+### Purpose &amp; impact 
+##### Background &amp; intent
+_Describe the desired state of the world after this project? Why does that matter?_
+<!--
+Outline the status quo, including any relevant context on the problem you’re seeing that this project should solve. Wherever possible, include pains or problems that you’ve seen users experience to help motivate why solving this problem works towards top-line objectives. 
+-->
+
+There is an inherent asymmetry in node capabilities imposed by the [constraints of the operating environment](#constraints-of-the-operating-environment). Attempts to overcome these limitations often lead to [ad-hoc stop-gap solutions that introduce centralization](#ad-hoc-stop-gap-solutions) into the system. Teams building products on IPFS are required to fill capability gaps with [(unintentionally) incompatible infrastructure](#Unintentionally-Incompatible-infrastructure) of their own.
+
+[IPFS HTTP API provides inadequate solution](#HTTP-API-provides-inadequate-solution) for constrained environments as it is designed to give a single client an **absolute control** of the host node as opposed to coordinating access of multiple tenants.
+
+Here we propose design which embraces asymmetry among node capabilities. Instead of traditional client-server model, symbiotic network is formed through a protocol that enables IPFS nodes with limited capabilities and/or lifespan _(from now on referred as **symbionts**)_ to partner with IPFS nodes that posses greater capabilities and/or lifespan _(from now on referred as **hosts**)_.
+
+<!-- In this design symbionts are a hybrid between IPFS Client and an IPFS node with limited capabilities _(Such as IPFS node embedded in a web site)_. In solitude symbionts operate in limited capacity. They can seamlessely upgrade thier capabilities through a partner host. -->
+
+Through the [symbiosis protocol](#Symbiosys-protocol) a **host** can announce [services](#Services) that **symbionts** can _request_ to overcome own limitations. 
+
+Protocol defines `m:n` relationship implying that a symbiont can have multiple partner hosts and a host could partner with multiple symbionts.
+
+
+##### Assumptions &amp; hypotheses
+_What must be true for this project to matter?_
+<!--(bullet list)-->
+
+- Neither embedded IPFS node nor IPFS client provide an adequate solution for building products in constrained environments.
+- OS level node sharing with existing building blocks is impractial.
+- Lack of generalized delegation protocol between nodes leads ad-hock stop-gap solutions.
+- Teams are unable to plug domain specific extensions without having to fork our stack.
```

Comment:
What type of domain specific extensions we are describing here? Dapp developers should create libp2p protocols

Reply:
There are bunch cases where domain specific replication strategy would be far more effective than bitswap is. In fact IPLD selectors is one example where that could be expressed a service. Another one I was thinking is textile threads, where communicating the a head you've seen is enough for other participants to know all the blocks they need to send you.

I think libp2p protocols are indeed relevant here. I think of this as a bit more higher level equivalent and designed with asymmetry in mind.